hal/arm/rtt.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Phoenix-RTOS
+ *
+ * SEGGER's Real Time Transfer - simplified driver
+ *
+ * Copyright 2023-2024 Phoenix Systems
+ * Author: Gerard Swiderski, Daniel Sawka
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "include/errno.h"
+#include "syspage.h"
+#include "hal/arm/barriers.h"
+
+#include <board_config.h>
+#include "rtt.h"
+
+#ifndef RTT_SYSPAGE_MAP_NAME
+#define RTT_SYSPAGE_MAP_NAME "rtt"
+#endif
+
+#ifndef RTT_CB_SIZE
+#define RTT_CB_SIZE 256
+#endif
+
+
+struct rtt_pipe {
+	const char *name;
+	volatile unsigned char *ptr;
+	unsigned int sz;
+	volatile unsigned int wr;
+	volatile unsigned int rd;
+	unsigned int flags;
+};
+
+
+struct rtt_desc {
+	char tag[16];
+	unsigned int txChannels;
+	unsigned int rxChannels;
+	struct rtt_pipe channels[];
+};
+
+
+static struct {
+	volatile struct rtt_desc *rtt;
+} common;
+
+
+int rtt_check(int chan, rtt_dir_t dir)
+{
+	if (chan < 0) {
+		return -EINVAL;
+	}
+
+	if (common.rtt == NULL) {
+		return -ENODEV;
+	}
+
+	if ((dir == rtt_dir_up) && (chan >= common.rtt->txChannels)) {
+		return -ENODEV;
+	}
+
+	if ((dir == rtt_dir_down) && (chan >= common.rtt->rxChannels)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+
+int _hal_rttRead(int chan, void *buf, unsigned int count)
+{
+	if (rtt_check(chan, rtt_dir_down) < 0) {
+		return -ENODEV;
+	}
+
+	volatile struct rtt_desc *rtt = common.rtt;
+
+	hal_cpuDataMemoryBarrier();
+	chan = rtt->txChannels + chan;
+	volatile unsigned char *srcBuf = rtt->channels[chan].ptr;
+	unsigned char *dstBuf = (unsigned char *)buf;
+	unsigned int sz = rtt->channels[chan].sz - 1;
+	unsigned int rd = rtt->channels[chan].rd & sz;
+	unsigned int wr = rtt->channels[chan].wr & sz;
+	unsigned int todo = count;
+
+	while ((todo != 0) && (rd != wr)) {
+		*dstBuf++ = srcBuf[rd];
+		rd = (rd + 1) & sz;
+		todo--;
+	}
+
+	hal_cpuDataMemoryBarrier();
+	rtt->channels[chan].rd = rd;
+
+	return count - todo;
+}
+
+
+int _hal_rttWrite(int chan, const void *buf, unsigned int count)
+{
+	if (rtt_check(chan, rtt_dir_up) < 0) {
+		return -ENODEV;
+	}
+
+	volatile struct rtt_desc *rtt = common.rtt;
+
+	hal_cpuDataMemoryBarrier();
+	const unsigned char *srcBuf = (const unsigned char *)buf;
+	volatile unsigned char *dstBuf = rtt->channels[chan].ptr;
+	unsigned int sz = rtt->channels[chan].sz - 1;
+	unsigned int rd = (rtt->channels[chan].rd + sz) & sz;
```

Comment:
This code will fail silently and do nothing when `count == N * sz`

Reply:
Let's assume that we start with rd == wr and count == sz. Then (rd + sz % sz) == rd and rd still == wr. We will not enter while below (&& (rd != wr)) and in consequence, do nothing at all

Reply:
You didn't notice that the variable `sz` is the buffer size minus 1. Therefore, the result of this line is equivalent to `(rtt->channels[chan].rd + rtt->channels[chan].sz - 1) % (rtt->channels[chan].sz)`

Reply:
Ok, then count == sz-1, problem persists ;)